synapse/handlers/sync.py

```diff
@@ -135,6 +138,12 @@ def wait_for_sync_for_user(self, sync_config, since_token=None, timeout=0,
             A Deferred SyncResult.
         """
 
+        if sync_config.is_guest:
+            for room_id in sync_config.filter.list_rooms():
+                world_readable = yield self._is_world_readable(room_id)
+                if not world_readable:
+                    raise AuthError(403, "Guest access not allowed")
```

Comment:
Yep. I made it match ``/events`` in that respect. Sadly I suspect that pointing out that you could bisect the list to search for the bad ids would not be considered constructive. I'll look at including the bad ids in the response somehow.